### PR TITLE
[Messenger] [FrameworkBundle] Added `semaphore` transport.

### PR DESCRIPTION
**Pull request by CedrickOka**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | any
| License       | MIT
| Doc PR        | any

# Semaphore Transport
The Semaphore Transport allows you to send and receive messages on System V message queues. The `semaphore` transport configuration looks like this :

```bash
# .env
MESSENGER_TRANSPORT_DSN=semaphore://%kernel.project_dir%/.env
# Full DSN Example
MESSENGER_TRANSPORT_DSN=semaphore://%kernel.project_dir%/.env?project=M&message_type=1&message_max_size=1024
```

A number of options can be configured via the DSN or via the options key under the transport in messenger.yaml:

Option | Description | Default 
-------- | --------------- | ---------
path | Pathname to create System V IPC key | 
project | Project identifier to create System V IPC key | M
message_type | The type of message to send | 1
message_max_size | The maximum size in bytes of a message if the message is larger than this size, an exception will be thrown | 131072
auto_setup | Enable or not the auto-setup of queue | true

> This extension is not available on Windows platforms.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+     *
+     * @throws SemaphoreException
+     */
+    public function send(string $body, array $headers = [], int $delay = 0, SemaphoreStamp $semaphoreStamp = null): void
+    {
+        $messageType = null !== $semaphoreStamp ? $semaphoreStamp->getType() : ($this->configuration['message_type'] ?? 1);
+        $message = json_encode(['body' => $body, 'headers' => $headers]);
+
+        if ($this->configuration['message_max_size'] < \strlen($message)) {
+            throw new SemaphoreException(sprintf('The semaphore message is too long to be sent, the maximum size accepted is 10 bytes.', $this->configuration['message_max_size']));
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]:
```diff
- 'The semaphore message is too long to be sent, the maximum size accepted is 10 bytes.'
+ 'The semaphore message is too long to be sent, the maximum size accepted is %s bytes.'
```

?

File: src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreReceiver.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            throw new TransportException($exception->getMessage(), 0, $exception);
+        }
+
+        if (null === $semaphoreEnvelope) {
+            return;
+        }
+
+        try {
+            $envelope = $this->serializer->decode([
+                    'body' => $semaphoreEnvelope->getBody(),
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreReceiver.php]: A bit too much indent? :)

File: src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreSender.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        $encodedMessage = $this->serializer->encode($envelope);
+
+        /** @var \Symfony\Component\Messenger\Stamp\DelayStamp|null $delayStamp */
+        $delayStamp = $envelope->last(DelayStamp::class);
+        $delay = null !== $delayStamp ? $delayStamp->getDelay() : 0;
+
+        try {
+            $this->connection->send(
+                    $encodedMessage['body'],
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreSender.php]: same here

File: src/Symfony/Component/Messenger/Transport/Semaphore/Util/PlatformUtil.php [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Semaphore\Util;
+
+/**
+ * @author Cedrick Oka Baidai <okacedrick@gmail.com>
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Util/PlatformUtil.php]: Should probably be marked `@internal`

Comment (ogizanagi) [reply src/Symfony/Component/Messenger/Transport/Semaphore/Util/PlatformUtil.php]: Btw, isn't `\DIRECTORY_SEPARATOR === '\\'` enough? I'm not sure we need a method for this.

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    private $queue;
+
+    /**
+     * @var bool
+     */
+    private $setup;
+
+    public function __construct(array $configuration)
+    {
+        if (true === PlatformUtil::isWindows()) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: Shouldn't we rather check `\extension_loaded('sysvsem')`, as for the Lock component `SemaphoreStore`?

https://github.com/symfony/symfony/blob/d2649f2f7845cf40dec7c0f44704382b99d30c56/src/Symfony/Component/Lock/Store/SemaphoreStore.php#L36-L41

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php (hunk 1 above)
Comment (CedrickOka) [reply src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: thanks for the mistake fixed

**Comment by CedrickOka** [comment]
Hi, 

@ogizanagi I applied your recommendations of your reviews but I have more news since.
I hope you are well.

File: src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreReceiver.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     * {@inheritdoc}
+     *
+     * @see \Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface::get()
+     */
+    public function get(): iterable
+    {
+        try {
+            $semaphoreEnvelope = $this->connection->get();
+        } catch (SemaphoreException $exception) {
+            throw new TransportException($exception->getMessage(), 0, $exception);
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreReceiver.php]: Same

File: src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreStamp.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+use Symfony\Component\Messenger\Stamp\NonSendableStampInterface;
+
+/**
+ * @author Cedrick Oka Baidai <okacedrick@gmail.com>
+ */
+class SemaphoreStamp implements NonSendableStampInterface
+{
+    /**
+     * @var int
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreStamp.php]: Such docblocks on properties can be removed as soon as the type is already identified by the constructor typehints & assigments

**Review by ogizanagi — COMMENTED** [review]
Code looks good and the feature promising to me so far, but I can't play with it for now.

**Comment by CedrickOka** [comment]
Hi @ogizanagi,

I applied your recommendations, thanks.

**Comment by CedrickOka** [comment]
Hi @ogizanagi,

I hope you're doing well, I'm still waiting for your reviews

**Comment by ogizanagi** [comment]
Hi @CedrickOka . As you fixed my review, code looks good to me, but I still need to find some time to give it a try (as anyone else could actually).

**Comment by CedrickOka** [comment]
> Hi @CedrickOka . As you fixed my review, code looks good to me, but I still need to find some time to give it a try (as anyone else could actually).

Thanks

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 8]
```diff
@@ -121,7 +118,7 @@
 use Symfony\Component\Validator\ObjectInitializerInterface;
 use Symfony\Component\Validator\Util\LegacyTranslatorProxy;
 use Symfony\Component\WebLink\HttpHeaderSerializer;
-use Symfony\Component\Workflow;
+use Symfony\Component\Workflow\Workflow;
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Same. This one should be removed in a dedicated PR on 4.4 instead.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 9]
```diff
@@ -42,6 +38,7 @@
 use Symfony\Component\Config\ResourceCheckerInterface;
 use Symfony\Component\Console\Application;
 use Symfony\Component\Console\Command\Command;
+use Symfony\Component\CssSelector\Parser\Reader;
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: These changes seems unrelated to this PR (and are wrong in the first place?)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 10]
```diff
@@ -12,11 +12,7 @@
 namespace Symfony\Bundle\FrameworkBundle\DependencyInjection;
 
 use Doctrine\Common\Annotations\AnnotationRegistry;
-use Doctrine\Common\Annotations\Reader;
-use Http\Client\HttpClient;
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Please revert

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 11]
```diff
@@ -12,11 +12,7 @@
...
 use Psr\Cache\CacheItemPoolInterface;
-use Psr\Container\ContainerInterface as PsrContainerInterface;
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Same

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 12]
```diff
@@ -64,6 +61,7 @@
 use Symfony\Component\Form\FormTypeExtensionInterface;
 use Symfony\Component\Form\FormTypeGuesserInterface;
 use Symfony\Component\Form\FormTypeInterface;
+use Symfony\Component\HttpClient\HttpClient;
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Same

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 13]
```diff
@@ -87,8 +85,7 @@
 use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
 use Symfony\Component\Messenger\MessageBus;
 use Symfony\Component\Messenger\MessageBusInterface;
-use Symfony\Component\Messenger\Transport\TransportFactoryInterface;
-use Symfony\Component\Messenger\Transport\TransportInterface;
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Same. Btw this one is currently the root cause of a bug preventing this PR to work.

**Review by ogizanagi — COMMENTED** [review]
So I gave it a try on an existing project. It seems to work great so far.

However, about using `MESSENGER_TRANSPORT_DSN=semaphore://%kernel.project_dir%/.env`
This means the env var must be resolve using `'%env(resolve:MESSENGER_TRANSPORT_DSN)%'` in order to get `%kernel.project_dir%` resolved, which is not the default in the recipe: https://github.com/symfony/recipes/blob/5f8acedd41d584611b2422aa3d4d66f97c70c3f2/symfony/messenger/4.3/config/packages/messenger.yaml#L8

**Comment by ogizanagi** [comment]
Also, this would now target 5.1.
Could you please rebase on master?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 14]
```diff
@@ -376,77 +375,77 @@ public function load(array $configs, ContainerBuilder $container)
         }
 
         $container->registerForAutoconfiguration(Command::class)
-            ->addTag('console.command');
+        ->addTag('console.command');
```
Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: This needs to be reverted

**Comment by Nyholm** [comment]
Thank you for this PR. I see that you done a great job with this, and this transport looks really good. 

However, I am strongly against adding new transports. Symfony should only provide transports that are use by the vast majority of people. AMQP, Redis, Doctrine, and custom transports like `sync://` and `in-memory://` is enough to cover most applications using messenger. 

I know there are special and super cool transports, but I believe that they should not live inside of Symfony organisation. 

I vote for closing this PR. And I strongly recommend @CedrickOka to create a library of this great code so other can enjoy this semaphore transport.

**Comment by Nyholm** [comment]
For the record: 
After discussions in the core team I've changed my mind. Transports like this should be in a separate package in the Symfony organization. Ie, not `symfony/messager`, but `symfony/semaphore-messenger`.

The PR will be merged if Semaphore is considered a popular transport.

**Comment by CedrickOka** [comment]
Hi @Nyholm,

Following your discussions, how should I proceed then or does the discussions continue?

**Comment by jderusse** [comment]
hi @chalasr AFAIK semaphore provides all features needed by messenger (non-blocking read, delivery once, buffering) I'm not against it (beware of limitation like maxsize, .

I personnaly use a real broker in order to scale consumers and probably won't use it.
But I see a real benefit for such transport: it's easy to install and deploy. It will be a perfect transport for Async Email.

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+     * Send message on semaphore queue.
+     *
+     * @param SemaphoreStamp $semaphoreStamp
+     *
+     * @throws SemaphoreException
+     */
+    public function send(string $body, array $headers = [], int $delay = 0, SemaphoreStamp $semaphoreStamp = null): void
+    {
+        $messageType = null !== $semaphoreStamp ? $semaphoreStamp->getType() : ($this->configuration['message_type'] ?? 1);
+        $message = json_encode(['body' => $body, 'headers' => $headers]);
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: assert json_encode didn't failed

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+        if ($this->configuration['message_max_size'] < \strlen($message)) {
+            throw new SemaphoreException(sprintf('The semaphore message is too long to be sent, the maximum size accepted is %s bytes.', $this->configuration['message_max_size']));
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: This is not related to Semaphore itself, is it?

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+            throw new SemaphoreException(sprintf('Semaphore sending message failed with error code : "%s".', $errorCode));
+        }
+    }
+
+    /**
+     * Waits and gets a message from the configured semaphore queue.
+     *
+     * @throws SemaphoreException
+     */
+    public function get(): ?SemaphoreEnvelope
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: `SemaphoreEnvelope` could be replaced by an array (see other transports)

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    public function getMessageCount(): int
+    {
+        $stat = msg_stat_queue($this->getQueue());
+
+        return $stat['msg_qnum'] ?? 0;
+    }
+
+    public function close(): void
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: not used

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    /**
+     * Creates a connection based on the DSN and options.
+     *
+     * Available options:
+     *
+     *   * path: Pathname to create System V IPC key (Default: __FILE__)
+     *   * project: Project identifier to create System V IPC key (Default: M)
+     *   * message_type: The type of message to send (Default: 1)
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: does those 2 parameter have to be customisalbe by the user? 
Isn't a unique path enougth to use this transport?

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    public function close(): void
+    {
+        msg_remove_queue($this->getQueue());
+    }
+
+    public function setup(): void
+    {
+        $key = ftok($this->configuration['path'], substr($this->configuration['project'], 0, 1));
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: should have thrown a LogicException if project is not a single char

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+        $this->queue = msg_get_queue($key);
+        $this->setup = true;
+    }
+
+    private function shouldSetup(): bool
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: this method could be replaced by a `auto_setup` property. Initialized in constructor see Redis for example

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+    {
+        if (true === $this->setup) {
+            return false;
+        }
+
+        if (true === \in_array($this->configuration['auto_setup'], [false, 'false'], true)) {
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: cast to bool should have been done in initialization

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * @return resource
+     */
+    private function getQueue()
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: no need for a private getter

File: src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+    {
+        if (true === $this->shouldSetup()) {
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php]: this breaks `SRP` (see RedisExt/Connection for example of setup)

File: src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreReceiver.php [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+            return;
+        }
+
+        try {
+            $envelope = $this->serializer->decode([
+                'body' => $semaphoreEnvelope->getBody(),
+                'headers' => $semaphoreEnvelope->getHeaders(),
+            ]);
+        } catch (MessageDecodingFailedException $exception) {
+            // TODO: [Researh] Implements nack strategy for semaphore
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Semaphore/SemaphoreReceiver.php]: remove this comment. 

I don't think it's possible. And I don't think you should nack here. If the message is not decodable once, it'll never be decodable

**Comment by jderusse** [comment]
hmmm.. just realized that semaphore missed a very important requirement...  It don't provides ack/reject behaviors. 

Meaning that, if the consumer stops in the middle of the process (unhandled exception, segfault, ...) the message will be lost.

That makes the transport not reliable. I'm not sure how important it is for you. But it will be a no-go for me. :(

**Comment by fabpot** [comment]
Based on the limitation listed by @jderusse, I don't think this is something we can do in core. Thanks for working on this PR.
